From 7a745292b503d4be5caf848f2aaeefef81d56772 Mon Sep 17 00:00:00 2001
From: Peter Marschall <peter@adpm.de>
Date: Fri, 15 Apr 2011 18:10:07 +0200
Subject: [PATCH 11/32] OpenPGP: update card capabilities from historical
 bytes

According to OpenPGP card specs 1.1 & 2.0 historical bytes in the ATR
indicate capabilities:
* bit 0x40 of the 3rd byte of the compact-TLV entry with TL 0x73 tells
  whether the card supports extended Lc/Le fields in APDUs.
---
 src/libopensc/card-openpgp.c |   15 +++++++++++++++
 1 files changed, 15 insertions(+), 0 deletions(-)

diff --git a/src/libopensc/card-openpgp.c b/src/libopensc/card-openpgp.c
index f81d943..8ba3475 100644
--- a/src/libopensc/card-openpgp.c
+++ b/src/libopensc/card-openpgp.c
@@ -207,6 +207,21 @@ pgp_init(sc_card_t *card)
 		return SC_ERROR_OUT_OF_MEMORY;
 	}
 
+	/* update card capabilities from ATR */
+	if (card->atr.len > 0) {
+		unsigned char *hist_bytes = card->atr.value;
+		size_t len = card->atr.len;
+		size_t i = 0;
+
+		while ((i < len) && (hist_bytes[i] != 0x73))
+			i++;
+
+		/* bit 0x40 in byte 3 of TL 0x73 means "extended Le/Lc" */
+		if ((hist_bytes[i] == 0x73) && (len > i+3) &&
+		    (hist_bytes[i+3] & 0x40))
+			card->caps |= SC_CARD_CAP_APDU_EXT;
+	}
+
 	return SC_SUCCESS;
 }
 
-- 
1.7.4.4

